be/src/runtime/sender_queue.h

```diff
@@ -0,0 +1,230 @@
+// This file is licensed under the Elastic License 2.0. Copyright 2021-present, StarRocks Limited.
+
+#pragma once
+#include "column/vectorized_fwd.h"
+#include "runtime/data_stream_recvr.h"
+#include "serde/protobuf_serde.h"
+#include "util/moodycamel/concurrentqueue.h"
+#include "util/spinlock.h"
+
+namespace google::protobuf {
+class Closure;
+} // namespace google::protobuf
+
+namespace starrocks {
+using vectorized::ChunkUniquePtr;
+
+class DataStreamRecvr::SenderQueue {
+public:
+    SenderQueue(DataStreamRecvr* parent_recvr) : _recvr(parent_recvr) {}
+    virtual ~SenderQueue() = default;
+
+    // get chunk from this sender queue, driver_sequence is only meaningful in pipeline engine
+    virtual Status get_chunk(vectorized::Chunk** chunk, const int32_t driver_sequence = -1) = 0;
+
+    // check if data has come, work with try_get_chunk.
+    virtual bool has_chunk() = 0;
+
+    // Probe for chunks, because _chunk_queue maybe empty when data hasn't come yet.
+    virtual bool try_get_chunk(vectorized::Chunk** chunk) = 0;
+
+    // add chunks to this sender queue if this stream has not been cancelled
+    virtual Status add_chunks(const PTransmitChunkParams& request, ::google::protobuf::Closure** done) = 0;
+
+    // add chunks to this sender queue if this stream has not been cancelled
+    // Process data in strict accordance with the order of the sequence
+    virtual Status add_chunks_and_keep_order(const PTransmitChunkParams& request,
+                                             ::google::protobuf::Closure** done) = 0;
+
+    // Decrement the number of remaining senders for this queue
+    virtual void decrement_senders(int be_number) = 0;
+
+    virtual void cancel() = 0;
+
+    virtual void close() = 0;
+
+protected:
+    Status _build_chunk_meta(const ChunkPB& pb_chunk);
+
+    Status _deserialize_chunk(const ChunkPB& pchunk, vectorized::Chunk* chunk, faststring* uncompressed_buffer);
+
+    DataStreamRecvr* _recvr = nullptr;
+
+    serde::ProtobufChunkMeta _chunk_meta;
+};
+
+class DataStreamRecvr::NonPipelineSenderQueue : public DataStreamRecvr::SenderQueue {
```

Comment:
Could the two derived class be moved to `.cpp` file? Nobody need to know the implementation but only the public interface?

Reply:
> Could the two derived class be moved to `.cpp` file? Nobody need to know the implementation but only the public interface?

`PipelineSenderQueue` has serveral interfaces that are not common, such as has_output and short_circuit, so the caller still needs to be aware of the definition…